[aot autograd] refactor to make functionalization self-contained

bdhirsh opened the pull request:
This refactor should make it easier to add an export hook into aot autograd.

(1) I killed `create_forward_or_joint_functionalized()` (and the functions that it called, like `forward_or_joint()`) which used to handle autograd + functionalization all-in-one-go for the joint case, and was also used in the inference case.

I added a few separate helper functions:

`create_functionalized_graph()`: this takes a flat fn, and returns a functionalized fx graph. It is mostly just a thin wrapper around functionalization + make_fx(), but also has some extra logic to manually append `copy_()` ops to the end of the graph.

`fn_no_extra_mutations()`: this creates the fn that we want to trace in the inference code path. It takes in a function that it then calls, and returns the outputs + any (updated) mutated inputs.

`joint_fn_no_external_mutations()`: this creates the fn that we want to trace in the joint code path. It takes in a function, and traces out its joint. It also does the work of cloning inputs that are mutated and require gradients, returning mutated inputs as outputs, and returning intermediate bases as outputs

We should be able to add an export hook by basically adding a similar version of `joint_fn_no_external_mutations` but with a lot more restrictions (guaranteed to have no tangents, not synthetic bases, etc), and calling `create_functionalized_graph()` on it.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96341

Differential Revision: [D44204090](https://our.internmc.facebook.com/intern/diff/D44204090)

bdhirsh committed:
[aot autograd] refactor to make functionalization self-contained

[ghstack-poisoned]

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -861,181 +876,145 @@ def maybe_to_fresh_input(idx, t, meta):
-#   except that every input we know will be mutated in the forward has been cloned.
-#   We run our forward on primals_after_cloning (potentially mutating some inputs), and then compute our gradients
-#   w.r.t. primals_before_cloning (so we properly capture the mutation in our gradient computation).
+def fn_no_extra_mutations(
```
comment? :)

bdhirsh replied:
adding

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -861,181 +876,145 @@ def maybe_to_fresh_input(idx, t, meta):
 ) -> Any:
-    outs = fn(*primals_after_cloning)
-    assert len(meta.output_info) == len(outs)
+    def inner_fn(primals: List[Any], tangents: List[Any]):
```
Get a better type here? Matching the docs

bdhirsh replied:
which docs are you referring to? I think we've always been doing List[Any], List[Any] to type the joint

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -861,181 +876,145 @@ def maybe_to_fresh_input(idx, t, meta):
-# - Our function will return (updated_inputs, fw_outs, intermediate_bases), (gradients)
-#
-# Finally, if keep_input_mutations is set, then we will explicitly *not* return updated inputs, for any inputs
+# Our function will return (updated_inputs, fw_outs, intermediate_bases), (gradients)
```
I'm not sure what this means

bdhirsh replied:
fair haha. I'll update the comment on my new `fn_prepped_for_autograd()` function

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -861,181 +876,145 @@ def maybe_to_fresh_input(idx, t, meta):
-#
-# Finally, if keep_input_mutations is set, then we will explicitly *not* return updated inputs, for any inputs
+# Our function will return (updated_inputs, fw_outs, intermediate_bases), (gradients)
+# If keep_input_mutations is set, then we will explicitly *not* return updated inputs, for any inputs
```
There is no argument named this though?

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -861,181 +876,145 @@ def maybe_to_fresh_input(idx, t, meta):
+def fn_no_extra_mutations(
+    fn: Callable,
+    meta: ViewAndMutationMeta,
+    keep_input_mutations: bool,
```
It's supremely unclear to me why we need all these knobs.

1. I still think we should have handled metadata mutation in its own pass early on
2. I'm not sure why we need `keep_input_mutations` to toggle what outputs we return. In particular, it seems harmless to unconditionally NOT return the mutated inputs and figure the copy_ calls later (because remember, you can always access the original inputs to get the mutated versions!)

bdhirsh replied:
From our last offline discussion:

(1) Agreed on handling metadata mutation earlier being nice (although there will be some mumbling around handling resize_(), and I probably won't deal with it in this PR)

(2) on the keep_input_mutation handling: it sounds like you're suggesting that all of the logic around "returning updated inputs as outputs" could be handled entirely by the functionalization layer. I'm not totally convinced we can do this: at some point when we call `autograd.grad()`, we need to potentially pass in the mutated inputs as `needed_outs` (and corresponding tangents for them), so that we properly autograd through the input mutation. This can't happen at the functionalization layer, because the fn that `create_functionalized_graph()` takes in is the entire joint function itself - it needs to be inside of that joint function that we pass mutated inputs into `autograd.grad`

Right now locally, I have a `prepare_for_autograd()` function that basically takes in a function, and does all of the intermediate work of (1) clone mutated inputs, (2) return mutated inputs, (3) return intermediate bases. I then pass that new function to a very simple `create_joint()` function, that runs the given fn and calls `autograd.grad()` on the output (along with a mask to know which outputs to grad).

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -861,181 +876,145 @@ def maybe_to_fresh_input(idx, t, meta):
+# Instead, we are relying on the logic in functionalize_fn to manually perform the input mutations,
 # keeping them directly in the traced graph.
-def forward_or_joint(
+def joint_fn_no_external_mutations(
```
What is meant by "no external mutations" here?

bdhirsh replied:
I'm going to rewrite this to have:

(1) `joint_fn()` be its own thing (that just runs function, and performs autograd.grad on the outputs that require gradients)

(2) a separate function that handles all of the input mutation / cloning / intermediate base handling. That's sort of what I meant by the "no_external_mutations" bit, although it's not a great name. For now locally, I named my function `new_fn = fn_prepped_for_autograd(fn, fw_metadata)`

ezyang reviewed (approved):
I'm approving this because it's an improvement. However, it feels like *half* of the refactor we're looking for. You've gotten the input convention nice and clean by the time you get to the inner function calls. But joint_fn_no_external_mutations still does a lot of querying on meta to setup `fw_outs_to_return`. But this seems distinctly not joint_fn_no_external_mutations's job: returning the mutated inputs ought to be create_functionalized_graph's job; after all, that's what actually does the functionalization! If you called these functions without create_functionalized_graph, it would be quite odd to see all of these seemingly useless outputs floating around.

bdhirsh committed:
Update on "[aot autograd] refactor to make functionalization self-contained"

This refactor should make it easier to add an export hook into aot autograd.

(1) I killed `create_forward_or_joint_functionalized()` (and the functions that it called, like `forward_or_joint()`) which used to handle autograd + functionalization all-in-one-go for the joint case, and was also used in the inference case.

I added a few separate helper functions:

`create_functionalized_graph()`: this takes a flat fn, and returns a functionalized fx graph. It is mostly just a thin wrapper around functionalization + make_fx(), but also has some extra logic to manually append `copy_()` ops to the end of the graph.

`fn_no_extra_mutations()`: this creates the fn that we want to trace in the inference code path. It takes in a function that it then calls, and returns the outputs + any (updated) mutated inputs.

`joint_fn_no_external_mutations()`: this creates the fn that we want to trace in the joint code path. It takes in a function, and traces out its joint. It also does the work of cloning inputs that are mutated and require gradients, returning mutated inputs as outputs, and returning intermediate bases as outputs

We should be able to add an export hook by basically adding a similar version of `joint_fn_no_external_mutations` but with a lot more restrictions (guaranteed to have no tangents, not synthetic bases, etc), and calling `create_functionalized_graph()` on it.




[ghstack-poisoned]

bdhirsh committed:
Update on "[aot autograd] refactor to make functionalization self-contained"

This refactor should make it easier to add an export hook into aot autograd.

(1) I killed `create_forward_or_joint_functionalized()` (and the functions that it called, like `forward_or_joint()`) which used to handle autograd + functionalization all-in-one-go for the joint case, and was also used in the inference case.

I added a few separate helper functions:

`create_functionalized_graph()`: this takes a flat fn, and returns a functionalized fx graph. It is mostly just a thin wrapper around functionalization + make_fx(), but also has some extra logic to manually append `copy_()` ops to the end of the graph.

`fn_no_extra_mutations()`: this creates the fn that we want to trace in the inference code path. It takes in a function that it then calls, and returns the outputs + any (updated) mutated inputs.

`joint_fn_no_external_mutations()`: this creates the fn that we want to trace in the joint code path. It takes in a function, and traces out its joint. It also does the work of cloning inputs that are mutated and require gradients, returning mutated inputs as outputs, and returning intermediate bases as outputs

We should be able to add an export hook by basically adding a similar version of `joint_fn_no_external_mutations` but with a lot more restrictions (guaranteed to have no tangents, not synthetic bases, etc), and calling `create_functionalized_graph()` on it.




[ghstack-poisoned]

DanilBaibak commented:
@pytorchbot revert -m "Break internal build" -c ghfirst

bdhirsh committed:
Update on "[aot autograd] refactor to make functionalization self-contained"

This refactor should make it easier to add an export hook into aot autograd.

(1) I killed `create_forward_or_joint_functionalized()` (and the functions that it called, like `forward_or_joint()`) which used to handle autograd + functionalization all-in-one-go for the joint case, and was also used in the inference case.

I added a few separate helper functions:

`create_functionalized_graph()`: this takes a flat fn, and returns a functionalized fx graph. It is mostly just a thin wrapper around functionalization + make_fx(), but also has some extra logic to manually append `copy_()` ops to the end of the graph.

`fn_no_extra_mutations()`: this creates the fn that we want to trace in the inference code path. It takes in a function that it then calls, and returns the outputs + any (updated) mutated inputs.

`joint_fn_no_external_mutations()`: this creates the fn that we want to trace in the joint code path. It takes in a function, and traces out its joint. It also does the work of cloning inputs that are mutated and require gradients, returning mutated inputs as outputs, and returning intermediate bases as outputs

We should be able to add an export hook by basically adding a similar version of `joint_fn_no_external_mutations` but with a lot more restrictions (guaranteed to have no tangents, not synthetic bases, etc), and calling `create_functionalized_graph()` on it.




[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44204090).

bdhirsh commented:
Confirmed that internal test passes

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "[aot autograd] refactor to make functionalization self-contained"

This refactor should make it easier to add an export hook into aot autograd.

(1) I killed `create_forward_or_joint_functionalized()` (and the functions that it called, like `forward_or_joint()`) which used to handle autograd + functionalization all-in-one-go for the joint case, and was also used in the inference case.

I added a few separate helper functions:

`create_functionalized_graph()`: this takes a flat fn, and returns a functionalized fx graph. It is mostly just a thin wrapper around functionalization + make_fx(), but also has some extra logic to manually append `copy_()` ops to the end of the graph.

`fn_no_extra_mutations()`: this creates the fn that we want to trace in the inference code path. It takes in a function that it then calls, and returns the outputs + any (updated) mutated inputs.

`joint_fn_no_external_mutations()`: this creates the fn that we want to trace in the joint code path. It takes in a function, and traces out its joint. It also does the work of cloning inputs that are mutated and require gradients, returning mutated inputs as outputs, and returning intermediate bases as outputs

We should be able to add an export hook by basically adding a similar version of `joint_fn_no_external_mutations` but with a lot more restrictions (guaranteed to have no tangents, not synthetic bases, etc), and calling `create_functionalized_graph()` on it.


Differential Revision: [D44204090](https://our.internmc.facebook.com/intern/diff/D44204090)

[ghstack-poisoned]

osalpekar commented:
@pytorchbot merge -f "Already landed internally as [D44204090](https://www.internalfb.com/diff/D44204090)"

bdhirsh committed:
Update on "[aot autograd] refactor to make functionalization self-contained"

This refactor should make it easier to add an export hook into aot autograd.

(1) I killed `create_forward_or_joint_functionalized()` (and the functions that it called, like `forward_or_joint()`) which used to handle autograd + functionalization all-in-one-go for the joint case, and was also used in the inference case.

I added a few separate helper functions:

`create_functionalized_graph()`: this takes a flat fn, and returns a functionalized fx graph. It is mostly just a thin wrapper around functionalization + make_fx(), but also has some extra logic to manually append `copy_()` ops to the end of the graph.

`fn_no_extra_mutations()`: this creates the fn that we want to trace in the inference code path. It takes in a function that it then calls, and returns the outputs + any (updated) mutated inputs.

`joint_fn_no_external_mutations()`: this creates the fn that we want to trace in the joint code path. It takes in a function, and traces out its joint. It also does the work of cloning inputs that are mutated and require gradients, returning mutated inputs as outputs, and returning intermediate bases as outputs

We should be able to add an export hook by basically adding a similar version of `joint_fn_no_external_mutations` but with a lot more restrictions (guaranteed to have no tangents, not synthetic bases, etc), and calling `create_functionalized_graph()` on it.


Differential Revision: [D44204090](https://our.internmc.facebook.com/intern/diff/D44204090)

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
pytorchbot merge -f "already landed internally"

bdhirsh commented:
@pytorchbot merge -f "already landed internally"
